lib/std/math/Interval.zig

```diff
@@ -0,0 +1,101 @@
+const std = @import("../std.zig");
+
+/// Create an interval of type T.
+/// See https://en.wikipedia.org/wiki/Interval_(mathematics) for more info.
+pub fn Interval(comptime T: type) type {
+    comptime std.debug.assert(std.meta.trait.isNumber(T)); // Interval must be instantiated with a number type.
+    return struct {
+        /// The left hand side of the Interval.
+        /// n as null here represents infinity.
+        /// open determines if the interval should be open (exclusive) or closed (inclusive). 
+        from: Endpoint,
+        /// The right hand side of the Interval.
+        /// n as null here represent infinity.
+        /// open determines if the interval should be open (exclusive) or closed (inclusive). 
+        to: Endpoint,
+
+        pub const Endpoint = struct {
+            n: ?T,
+            open: bool,
+        };
```

Comment:
It seems to me that the optional modifier should be put on the fields `from` and `to` rather than `n`, since I'm not sure that the "open" field means anything when `n` is null.

Or you could use a tagged union like this:

```zig
const Endpoint = union(enum) {
    Infinity: void,
    Bounded: struct { val: T, open: bool },
};
```

This is another variation that could be used.
```zig
const Endpoint = union(enum) {
    Infinity: void,
    Open: T,
    Closed: T,
};
```

Here's what it initializing a field would look like using the various representations:
```zig
// current
const unbounded = Interval(u8) { .from = .{ .n = null, .open = undefined }, .to = .{ .n = null, .open = undefined } };
// optional field
const  unbounded = Interval(u8) { .from = null, .to = null };
// tagged union
const unbounded = Interval(u8) { .from = EndPoint.Infinity, .to = EndPoint.Infinity };
```

Reply:
I like this a lot, I'll update

Reply:
updated 👍